Add scip::ObjExprhdlr class to objscip

kkofler opened the pull request:
The intent is to be able to use SWIG director classes to implement expression handlers through bindings, in particular, to add support for expression handlers to JSCIP. That needs the C++ class to build upon.

src/objscip/objexprhdlr.h, src/objscip/objexprhdlr.cpp: New files. C++ wrapper for expression handlers. The files contain the scip::ObjExprhdlr class and the functions SCIPincludeObjExprhdlr, SCIPfindObjExprhdlr, and SCIPgetObjExprhdlr.

src/CMakeLists.txt (objscipsources): Add objscip/objexprhdlr.cpp.
(objscipheaders): Add objscip/objexprhdlr.h.

svigerske commented on src/objscip/objexprhdlr.h:
```diff
@@ -496,7 +494,7 @@ class ObjExprhdlr : public ObjProbCloneable
 SCIP_EXPORT
 SCIP_RETCODE SCIPincludeObjExprhdlr(
    SCIP*                 scip,               /**< SCIP data structure */
-   SCIP_EXPRHDLR**       exprhdlr,           /**< pointer to store the expression handler */   scip::ObjExprhdlr*    objconshdlr,        /**< expression handler object */
```
I was aligning it with the include functions I saw for the other C++ plugins.
The idea seems to be that the user calls `SCIPfindExprhdlr()` to get the C version.

If it is important for you to get the `SCIP_EXPRHDLR`, then we could add it as optional last argument to `SCIPincludeObjExprhdlr`.

svigerske commented:
@kkofler Are you ok with the current changes to your changes?

svigerske commented:
Thank you for the contribution!

svigerske commented:
> The Java ExpressionHandler class wants to call SCIPincludeObjExprhdlr in the constructor (in any constructor,

Could you remove that it calls `SCIPincludeObjExprhdlr()` in any constructor?

Looking at objconshdlr and ConshdlrSubtour in the TSP example, there are not that many calls to `SCIPincludeObjConshdlr`. It's in `conshdlrCopyObj` only.

kkofler commented:
> Looking at objconshdlr and ConshdlrSubtour in the TSP example, there are not that many calls to `SCIPincludeObjConshdlr`. It's in `conshdlrCopyObj` only.

It is actually also called in the TSP example's `cppmain.cpp` in `runSCIP`. Basically, it must be called from somewhere so that the constraint handler can actually be used.

That said:

> Could you remove that it calls `SCIPincludeObjExprhdlr()` in any constructor?

I originally thought that would not be feasible, but I now think I have a way to do that in the Java binding without needing 2 separate `ExpressionHandler` wrapper classes (one for the C++ `scip::ObjExprHdlr*` and one for the C `SCIP_EXPRHDLR*`): Basically, I can leave the `ExpressionHandler`'s internal `SCIP_EXPRHDLR*` initially set to `nullptr` (actually, since the wrapper class `ExpressionHandler` is in Java, the internal `SWIGTYPE_p_SCIP_Exprhdlr` set to `null`) and let the `SCIPincludeObjExprhdlr` wrapper method set it. Anything trying to use the `SCIP_EXPRHDLR` getters (once I actually wrap them in the Java class, which is not even implemented yet) before the pointer is set will then fail, but that is expected. (The getters will also fail on the cloned `ExpressionHandler` class from the `exprhdlrCopyhdlr` operation, but nobody should be running the getters on that one from Java anyway. The getters only provide statistics and are not useful to implement the `ExpressionHandler` callback methods.)

That does mean, though, that a `SCIPincludeObjExprhdlr` equivalent has to be explicitly called, whereas the current Java bindings do it automatically (in the constructor, which needs the `Scip` object anyway). Though that matches how the C and C++ interfaces work.

kkofler commented:
> So couldn't the Java wrapper also use `SCIPfindExprhdlr()` to find the `SCIP_EXPRHDLR*`?

That sounds workable, yes. I am giving that a try now.

kkofler commented:
This is now implemented in JSCIPOpt (`exprhdlr` branch):
* `ExpressionHandler` constructors no longer automatically call `SCIPincludeObjExprhdlr`. Instead, there is an `include()` method that must be called explicitly.
* The internal `SCIP_EXPRHDLR*` is set by `include()`. Before, it is null. Methods that need the pointer now all use the `getPtr()` getter instead of accessing the field directly. If `getPtr()` is called with a null `SCIP_EXPRHDLR*`, it will try to look up the pointer using `SCIPfindExprhdlr`, in case `SCIPincludeObjExprhdlr` was called from C++. If that also returns null, an exception is thrown, telling the user that `include()` must be called first. (I could try to be extra smart and automatically call `include()` in that case, so `include()` would normally end up getting called when the first expression using the expression handler is constructed, but I am currently not convinced that that kind of automagic is more helpful than confusing.)
* The proposed patch from PR #159 is obsolete and no longer needed. I have also dropped it from my `v92-objexprhdlr-backport` branch that I use for testing.
